readchar/readchar_linux.py

```diff
@@ -1,18 +1,78 @@
 # -*- coding: utf-8 -*-
-# Initially taken from:
-# http://code.activestate.com/recipes/134892/
-# Thanks to Danny Yoo
+# Copyright (c) 2014 - 2017 Miguel Ángel García (@magmax9).
+# Licensed under the MIT license.
+# This code based on code snippet from the python 2 docs:
+# https://docs.python.org/2/faq/library.html
+#     #how-do-i-get-a-single-keypress-at-a-time
+
+
 import sys
-import tty
+import os
 import termios
+import fcntl
+from . import key
+
+
+def get_char(test_stream=None):
+    charbuffer = ''
+    readstream = read_char
+
+    if test_stream:
+        def altreader(*args):
+            return test_stream()
+        readstream = altreader
```

Comment:
get_character must return the test_stream.  To be truthful, I don't like any of this.  It is a hack purely implemented as a "work around" due to the Travis test suite (test steam is the fake data travis tests with).  I would like to eliminate ALL of this and just use read_char always (no teststream or altreaders).

It is implemented this way because if there is a test_stream, it just echos the steam back instead of processing it.  This is because the way readchar_linux is now implemented is not compatible with your factory test methods in Travis.